MudSwipeArea: Add option to prevent default on touch events.

esso23 opened the pull request:
## Description

Fixes #3047 

## How Has This Been Tested?
No tests added.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

esso23 committed:
MudSwipeArea: Add option to prevent default browser behavior on touch events.

esso23 committed:
MudSwipeArea: Codestyle.

esso23 committed:
mudHelpers: Codestyle.

esso23 commented:
Is there anything else I need to do to get this approved? I can't see what the error from Codacy is, it seems to be broken. But I think it was some unused variable, which was wrongly evaluated.

JonBunator commented:
What do you think @henon ? I think javascript is necessary here

henon commented on src/MudBlazor/Components/SwipeArea/MudSwipeArea.razor.cs:
```diff
@@ -1,19 +1,37 @@
         [Parameter]
         public Action<SwipeDirection> OnSwipe { get; set; }
 
+        [Parameter]
+        public bool PreventDefault { get; set; }
```
Please add an xmldoc comment to the property, explaining it.

henon commented:
A few things.

Doesn't this create a tiny little memory leak on the JS side when the component is disposed? I guess the touchHandlers should be unlinked on dispose. And what if multiple swipe areas register their JS handler over each other in browser window? Wouldn't the GC throw the handler of all the other SwipeAreas away and only the last would work as intended?

We have a policy to envelop all JS usage in either extension methods of ElementReference or in a blazor service.  The clean way would be to make a little service with a `Register` method which locks when registering the prevention handler so that there is no race condition between multiple components, increments an internal reference counter (also inside the lock) and decrements the counter when components Unregister when they are disposed. When the counter is 0 the handler can be nulled out in the window.

We have many services that work like this (some are a bit more complex than what I described above) like ResizeListener, KeyInterceptor, etc. so you can take some inspiration. The JS is in folder `TScript`, the service in folder `Services`.

henon commented:
> Doesn't this create a tiny little memory leak on the JS side when the component is disposed? I guess the touchHandlers should be unlinked on dispose. 

Sorry, no memleak, actually, they would be GCed with the element ref.

esso23 committed:
Merge remote-tracking branch 'mudblazor-origin/dev' into issue-3047

# Conflicts:
#	src/MudBlazor/Components/SwipeArea/MudSwipeArea.razor.cs

esso23 committed:
MudSwipeArea: Add Category attr and xmldoc for PreventDefault parameter.

henon commented:
@just-the-benno I reviewed this some time ago. Would love your opinion also

esso23 committed:
Remove .orig file.

esso23 committed:
SwipeArea: Add documentation for PreventDefault parameter.

esso23 commented:
Sorry for the delay. Let me know if any more edits are needed.

just-the-benno commented:
I promise to review this PR next.

just-the-benno commented on src/MudBlazor.Docs/Pages/Components/SwipeArea/Examples/SwipeDirectionsPreventDefaultExample.razor:
```diff
@@ -0,0 +1,7 @@
+﻿@namespace MudBlazor.Docs.Examples
+
+<MudSwipeArea OnSwipe="@((d) => _swipeDirection = d)" Style="width: 100%; height: 300px" PreventDefault>
+    <MudText Typo="@Typo.body1">@_swipeDirection</MudText>
```
I'd add a switch to toggle the value of PreventDefault so that user can see the effect directly

just-the-benno commented on src/MudBlazor/Components/SwipeArea/MudSwipeArea.razor.cs:
```diff
@@ -16,6 +21,24 @@ public partial class MudSwipeArea : MudComponentBase
+                await JsRuntime.InvokeVoidAsync("addDefaultPreventingHandler", _componentRef, "touchstart");
+                await JsRuntime.InvokeVoidAsync("addDefaultPreventingHandler", _componentRef, "touchend");
+                await JsRuntime.InvokeVoidAsync("addDefaultPreventingHandler", _componentRef, "touchcancel");
+            }
```
These are three calls to the Jsinterop service, so for Blazor Server, three calls must be awaited and are delayed. Could you rewrite it as one call? Maybe you can put the events as ``param`` or array argument?  This would include changes to js as well. 

Besides, I guess your methods are a good fit for the ``ElementReferenceExtensions`` class. (https://github.com/MudBlazor/MudBlazor/blob/dev/src/MudBlazor/Extensions/ElementReferenceExtensions.cs)

just-the-benno commented on src/MudBlazor/Components/SwipeArea/MudSwipeArea.razor.cs:
```diff
@@ -16,6 +21,24 @@ public partial class MudSwipeArea : MudComponentBase
+        protected override async Task OnAfterRenderAsync(bool firstRender)
+        {
+            if (firstRender && PreventDefault)
+            {
```
With this approach, you assume that the value of ``PreventDefault`` won't change over the lifetime of a ``MudSwipArea``. I guess this is true for most cases, but it'd be great to have the flexibility. This also included removing the handler dynamically. 

It is a good practices to have the call to the js interop layer within ``OnAfterRenderAsync``, however, I'd use the ``OnParametersSet`` method to create a flag, that is read in ``OnAfterRenderAsync` and - if needed - makes the js calls.

just-the-benno reviewed (changes requested):
Hi @esso23 

First of all, I want to apologize that we haven't kept up with this PR.  I think you are on a good path. 

I made some comments and suggestions. What is stopping it from merging is that I'd like to see that ``PreventDefaults`` can change over the life of a ``MudSwipeArea`` and also at least one unit test.

I try my best to be responsive (at least one day delay) in case you have questions etc.

esso23 committed:
SwipeArea: Use one call to register all handlers.

- Move addDefaultPreventingHandler to ElementReferenceExtensions
- Also implement addDefaultPreventingHandlers to register multiple handlers in one call

esso23 committed:
SwipeArea: Add support for PreventDefault changes during component lifetime.

esso23 committed:
SwipeArea/Docs: Add switch for PreventDefault.

esso23 committed:
SwipeArea: Remove references to listeners when they are removed in JS.

esso23 committed:
SwipeArea: Add tests for PreventDefault switching.

just-the-benno commented on src/MudBlazor/Components/SwipeArea/MudSwipeArea.razor.cs:
```diff
@@ -1,12 +1,19 @@
+        private static readonly string[] preventDefaultEventNames = { "touchstart", "touchend", "touchcancel" };
+        internal int[] _listenerIds;
+
+        [Inject] public IJSRuntime JsRuntime { get; set; }
```
Since you changed your JS calls to the extension methods (thanks for that), this reference - I guess - is not needed anymore.

just-the-benno commented on src/MudBlazor/Components/SwipeArea/MudSwipeArea.razor.cs:
```diff
@@ -1,12 +1,19 @@
     {
         internal double? _xDown, _yDown;
+        internal ElementReference _componentRef;
+        private static readonly string[] preventDefaultEventNames = { "touchstart", "touchend", "touchcancel" };
```
perfect! And thanks again for that change. Can you prepend it with an ``_`` as the other private fields as well?

esso23 replied:
Is the convention really the same for static and non-static fields? Is there some doc for naming conventions in MudBlazor?

henon replied:
yes, they are treated the same, both with leading _ following a lowercase letter. The only document we have that documents the conventions is the .editorconfig file ;)

henon replied:
btw, for the most part we are trying to follow the same guidlines as the .net foundation. but I don't know exactly how they treat private static fields. we just did it like this and now it has become convention.

just-the-benno reviewed (changes requested):
Hi @esso23, and thank you so much for incorporating the feedback. Only two minor things are left (see comments), and we are good to go.

esso23 committed:
SwipeArea: Remove unused property.

esso23 committed:
SwipeArea: Field naming convention.

esso23 committed:
fixup! SwipeArea: Field naming convention.

-_-

just-the-benno reviewed (approved):
Hi @esso23 

Thanks again for the changes, and I'm sorry for the delay. But... we are ready to merge it

just-the-benno commented:
@henon Ready for your review

henon commented:
Thanks @esso23 !
